consensus/helpers.go

```diff
@@ -137,35 +137,41 @@ func protoHash(m proto.Message) string {
 
 /*** P2P Helpers ***/
 
-func (m *consensusModule) sendToNode(msg *typesCons.HotstuffMessage) {
+func (m *consensusModule) sendToLeader(msg *typesCons.HotstuffMessage) {
+	m.nodeLog(typesCons.SendingMessage(msg, *m.leaderId))
+
 	// TODO(olshansky): This can happen due to a race condition with the pacemaker.
 	if m.leaderId == nil {
 		m.nodeLogError(typesCons.ErrNilLeaderId.Error(), nil)
 		return
 	}
 
-	m.nodeLog(typesCons.SendingMessage(msg, *m.leaderId))
 	anyConsensusMessage, err := codec.GetCodec().ToAny(msg)
 	if err != nil {
 		m.nodeLogError(typesCons.ErrCreateConsensusMessage.Error(), err)
 		return
 	}
+
 	if err := m.GetBus().GetP2PModule().Send(cryptoPocket.AddressFromString(m.idToValAddrMap[*m.leaderId]), anyConsensusMessage); err != nil {
 		m.nodeLogError(typesCons.ErrSendMessage.Error(), err)
 		return
 	}
 }
 
-func (m *consensusModule) broadcastToNodes(msg *typesCons.HotstuffMessage) {
+// A start-pattern broadcast (i.e. leader to all other validators) broadcast
+func (m *consensusModule) broadcastToValidators(msg *typesCons.HotstuffMessage) {
 	m.nodeLog(typesCons.BroadcastingMessage(msg))
+
 	anyConsensusMessage, err := codec.GetCodec().ToAny(msg)
 	if err != nil {
 		m.nodeLogError(typesCons.ErrCreateConsensusMessage.Error(), err)
 		return
 	}
-	if err := m.GetBus().GetP2PModule().Broadcast(anyConsensusMessage); err != nil {
-		m.nodeLogError(typesCons.ErrBroadcastMessage.Error(), err)
-		return
+
+	for _, val := range m.validatorMap {
+		if err := m.GetBus().GetP2PModule().Send(cryptoPocket.AddressFromString(val.GetAddress()), anyConsensusMessage); err != nil {
+			m.nodeLogError(typesCons.ErrBroadcastMessage.Error(), err)
+		}
```

Comment:
I am not sure I like this... we have P2P concerns leaking into the consensus module. Do we want that?

Reply:
Also https://github.com/pokt-network/pocket-network-protocol/blob/main/p2p/README.md#i-requirements

![image](https://user-images.githubusercontent.com/29378614/208077036-8ec45ca6-92f2-4b3e-ad60-ba22e93b9f39.png)

If this is a `HACK` I guess it should be written there otherwise the reader could get very confused very quickly.

Reply:
It's not a hack, but rather a previous incorrect implementation. See the details from the github message:

* `BroadcastToValidators` - Direct `O(N)` send from a leader to all validator replicas used by HotPOKT
* `BroadcastToStakedActors` - `O(log3N)` broadcast using RainTree from an originator to all staked nodes
* `RandomBroadcast` - Random gossip to all nodes (staked or not), including full nodes

[This article](https://medium.com/ontologynetwork/hotstuff-the-consensus-protocol-behind-facebooks-librabft-a5503680b151) has a very nice visual for basic hotstuff:

<img width="837" alt="Screenshot 2022-12-16 at 2 04 39 PM" src="https://user-images.githubusercontent.com/1892194/208196611-2b634cb8-975e-46a2-9927-73ae5dbf2f7c.png">

- During consensus: the validators use star pattern
- During tx & block propagation: staked actors use structured gossip
- During state sync and block propagation with full nodes: random gossip

Does that make sense?